projects/feray/feray/service.py

```diff
@@ -0,0 +1,270 @@
+    ):
+        """Deletes data for a given entity, with optional cascading to downstream assets."""
+        self.logger.info(
+            "\n--- Deleting data for entity %s starting from '%s' ---",
+            entity_key_values,
+            start_asset,
+        )
+
+        # --- THE FIX: Create a canonical hash string that matches the one in polars_add_row_id_fn ---
+        definition = self.registry.definitions[start_asset]
+        try:
+            # Build the string by joining the entity key values in their defined order.
+            canonical_string = "".join(
+                [str(entity_key_values[k]) for k in definition.entity_keys]
+            )
+        except KeyError as e:
+            raise ValueError(
+                f"Missing entity key {e} in provided values for asset '{start_asset}'."
+            ) from e
+
+        row_id_to_delete = RowId(hashlib.sha256(canonical_string.encode()).hexdigest())
+
+        assets_to_process = self._resolve_cascading_assets(start_asset, cascade)
+        self.logger.info("Deletion will apply to assets: %s", assets_to_process)
+
+        for asset_key in assets_to_process:
+            self._fs.delete_ids(asset_key=asset_key, row_ids=[row_id_to_delete])
+
+    def prune_asset_versions(
+        self,
+        *,
+        asset_key: S,
+        keep_last_n: int | None = None,
+        max_age: timedelta | None = None,
+    ):
+        """Prunes old data and metadata versions for a given asset based on policies."""
+        self.logger.info("\n--- Pruning versions for asset '%s' ---", asset_key)
+        all_prov = self._ms.all_provenance(asset_key=asset_key)
+        if not all_prov:
+            return
+
+        provs_to_delete: list[Provenance] = []
+        if keep_last_n is not None and len(all_prov) > keep_last_n:
+            provs_to_delete = all_prov[:-keep_last_n]
+
+        if max_age is not None:
+            self.logger.warning("WARNING: Pruning by max_age is not fully implemented.")
+
+        if provs_to_delete:
+            storage_ids_to_delete = [StorageId(self.hasher(p)) for p in provs_to_delete]
+            self.logger.info(
+                "  - Pruning %d versions from stores.", len(storage_ids_to_delete)
+            )
+            self._fs.prune_data_versions(
+                asset_key=asset_key, storage_ids_to_delete=storage_ids_to_delete
+            )
+            self._ms.prune_provenance(
+                asset_key=asset_key, storage_ids_to_delete=storage_ids_to_delete
+            )
+
+    def _materialize_one(self, asset_key: S, force_recompute: bool) -> DataFrameType:
+        """The core logic for materializing a single asset."""
+        definition = self.registry.definitions[asset_key]
+        latest_stored_prov = self._ms.latest_provenance(asset_key=asset_key)
+
+        latest_stored_storage_id: StorageId | None = (
+            StorageId(self.hasher(latest_stored_prov)) if latest_stored_prov else None
```

Comment:
nitpick: having a small `self._get_storage_id(provenance: Provenance)` helper method to call in these cases might be a nice way to abstract this repeated `StorageId(self.hasher(prov))` usage. Or maybe that logic could get pushed into the `Provenance` class, so you could do `self.hasher.id_for_provenance(provenance: Provenance)`.